File: TD2/Typescript/session2/td/natParInt.ts
import {
    Nat, FabriqueNat
} from "./naturels";

export class NatParInt implements Nat {

    constructor(private valeur: number) {
    }

    creerNatAvecValeur(valeur: number): Nat {
        if (valeur < 0) {
            throw new Error('* Erreur : valeur négative.');
        }
        if (valeur % 1 !== 0) {
            throw new Error('* Erreur : valeur non entière.');
        }
        return new NatParInt(valeur);
    }

    creerZero(): Nat {
        return this.creerNatAvecValeur(0);
    }

    creerSuccesseur(predecesseur: Nat): Nat {
        return this.creerNatAvecValeur(predecesseur.val() + 1);
    }

    creerNatAvecRepresentation(repDecimale: string): Nat {
        return this.creerNatAvecValeur(parseInt(repDecimale));
    }


    val(): number {
        return this.valeur;
    }

    estNul(): boolean {
        return this.val() === 0;
    }

    predecesseur(): Nat {
        if (this.estNul()) {
            throw new Error("* Erreur : naturel nul sans prédécesseur.");
        }
        return this.creerNatAvecValeur(this.val() - 1);
    }

    taille(): number {
        return this.representation().length;
    }

    chiffre(i: number): number {
        return parseInt(this.representation().charAt(this.taille() - 1 - i));
    }


    somme(x: Nat): Nat {
        return this.creerNatAvecValeur(x.val() + this.val());
    }

    zero(): Nat {
        return this.creerZero();
    }

    produit(x: Nat): Nat {
        return this.creerNatAvecValeur(x.val() * this.val());
    }

    un(): Nat {
        return this.creerNatAvecValeur(1);
    }

    modulo(x: Nat): Nat {
        return this.creerNatAvecValeur(this.val() % x.val());
    }

    div(x: Nat): Nat {
        return this.creerNatAvecValeur(Math.floor(this.val() / x.val()));
    }

    representation(): string {
        return this.val().toString();
    }

    estEgal(n: Nat): boolean {
        return n.val() == this.val();
    }


}

export const natParInt: FabriqueNat<Nat> = new NatParInt(0);

File: Java/src/session2/td/heritage/NatParInt.java
package session2.td.heritage;

import session2.td.naturels.AlgebreNatParInt;
import session2.td.naturels.Nat;

public class NatParInt extends IntPositif implements AlgebreNatParInt {
	
	public NatParInt(int val) {
		super(val);
	}
	
	public Nat creerNatAvecValeur(int val) {
		if(val==0) {
			return this.creerZero();
		} else {
			return new NatParInt(val);
		}
	}
	
	public String toString() {
		return Integer.toString(this.val());
	}


	public boolean equals(Object obj) {
		if(!(obj instanceof Nat))
			return false;
		Nat x = (Nat)obj;
		return this.val() == x.val();
	}

	@Override
	public Nat creerSuccesseur(Nat predecesseur) {
		return new NatParInt(predecesseur.val() +1);
	}

//	@Override
//	public String representation() {
//		return this.toString();
//	}
//	
}

